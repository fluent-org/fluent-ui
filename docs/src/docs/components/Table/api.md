---
title: Table
components: Table
api: true
langKey: "en"
---

## API

```
import Table from '@fluent-ui/core/Table'
import { TableGroup } from '@fluent-ui/core/Table'
// or
import { Table, TableGroup } from '@fluent-ui/core'
```

### Props

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| children | React.ReactElement &or; React.ReactElement[] |  | Display the content without `columns`. |
| data | any[] |  | Data of the table. |
| columns | ColumnsType[] |  | If you specify `columns`, the content of `children` will not be displayed. Please see below. |

#### ColumnsType

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| field&nbsp;* | string |  | Uniquely identifies. |
| title&nbsp;* | string |  | Title of the table. |
| render | (text: React.ReactText, record?: T, index?: number) => React.ReactElement |  | Render function for complex situations, which will be applied to the cell corresponding to each `field`(The parameters are expressed as: current cell value, current row data, current row index). |

#### UsedInTypeScript

It's worth noting that in order to support typescript more friendly, I gave up `React.forwardRef`, which means that using `ref` for the `Table` won't get what you want.

```tsx
import { ColumnsType, DataType } from '@fluent-ui/core/Table'

interface Data extends DataType {
  name: React.ReactText
  surname: React.ReactText
  birthYear: React.ReactText
  birthCity: React.ReactText
}

const data: Data[] = [
  { key: '1', name: 'Mehmet', surname: 'Baran', birthYear: 1987, birthCity: 22 },
  { key: '2', name: 'John', surname: 'Brown', birthYear: 1977, birthCity: 32 },
  { key: '3', name: 'Jim', surname: 'Black', birthYear: 1997, birthCity: 14 }
]

const columns: ColumnsType<Data>[] = [
  {
    title: 'Name',
    field: 'name',
    render: (text: React.ReactText): React.ReactElement => <a href="#">{text}</a> // eslint-disable-line
  },
  { title: 'Surname', field: 'surname' },
  { title: 'BirthYear', field: 'birthYear' },
  {
    title: 'BirthCity',
    field: 'birthCity'
  }
]

<Table<Data> columns={columns} data={data} />
```
