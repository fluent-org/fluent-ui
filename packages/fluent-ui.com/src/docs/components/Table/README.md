---
title: Table
components: Table
type: DataDisplay
---

# Table

<p class="description">Data tables display sets of data.</p>

## Simple Table

```jsx
() => {
  const data = [
    { key: '1', name: 'Mehmet', surname: 'Baran', birthYear: 1987, birthCity: 22 },
    { key: '2', name: 'John', surname: 'Brown', birthYear: 1977, birthCity: 32 },
    { key: '3', name: 'Jim', surname: 'Black', birthYear: 1997, birthCity: 14 }
  ]
  return (
    <Table>
      <Table.Head>
        <Table.Row>
          <Table.Cell>Name</Table.Cell>
          <Table.Cell>Surname</Table.Cell>
          <Table.Cell>BirthYear</Table.Cell>
          <Table.Cell>BirthCity</Table.Cell>
        </Table.Row>
      </Table.Head>
      <Table.Body>
        {data.map(row => (
          <Table.Row key={row.key}>
            <Table.Cell>{row.name}</Table.Cell>
            <Table.Cell>{row.surname}</Table.Cell>
            <Table.Cell>{row.birthYear}</Table.Cell>
            <Table.Cell>{row.birthCity}</Table.Cell>
          </Table.Row>
        ))}
      </Table.Body>
    </Table>
  )
}
```

## With Columns Table

```jsx
() => {
  const columns = [
    {
      title: 'Name',
      field: 'name',
      render: text => <a href="#">{text}</a>
    },
    { title: 'Surname', field: 'surname' },
    { title: 'BirthYear', field: 'birthYear' },
    {
      title: 'BirthCity',
      field: 'birthCity',
    }
  ]
  const data = [
    { key: '1', name: 'Mehmet', surname: 'Baran', birthYear: 1987, birthCity: 22 },
    { key: '2', name: 'John', surname: 'Brown', birthYear: 1977, birthCity: 32 },
    { key: '3', name: 'Jim', surname: 'Black', birthYear: 1997, birthCity: 14 }
  ]
  return <Table columns={columns} data={data} />
}
```
