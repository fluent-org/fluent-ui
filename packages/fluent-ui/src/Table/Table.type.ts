import * as React from 'react'
import * as PropTypes from 'prop-types'
import { StandardProps } from '..'

export interface DataType {
  key: string
  [name: string]: React.ReactText
}

export interface ColumnsType<T> {
  field: string
  title: string
  render?: (text: React.ReactText, record?: T, index?: number) => React.ReactElement
}

export interface TableProps<T>
  extends StandardProps<React.TableHTMLAttributes<HTMLTableElement>, HTMLTableElement> {
  data?: T[]
  columns?: ColumnsType<T>[]
  children?: React.ReactElement | React.ReactElement[]
}

export type TextAlign = 'inherit' | 'left' | 'center' | 'right' | 'justify'
export interface TableCellProps {
  textAlign?: TextAlign
  children: React.ReactChild
}

export const TablePropTypes = {
  data: PropTypes.array,
  columns: PropTypes.array,
  children: PropTypes.oneOfType([
    PropTypes.element,
    // @ts-ignore
    PropTypes.arrayOf<PropTypes.Requireable<PropTypes.ReactElementLike>>(PropTypes.element)
  ])
}
