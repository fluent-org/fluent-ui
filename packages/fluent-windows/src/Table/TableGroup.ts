import Head from './components/Head'
import Body from './components/Body'
import Row from './components/Row'
import Cell from './components/Cell'
import Footer from './components/Footer'

export const TableGroup = {
  Head,
  Body,
  Row,
  Cell,
  Footer
}
