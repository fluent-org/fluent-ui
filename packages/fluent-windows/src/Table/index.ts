export { default } from './Table'
export * from './TableGroup'
export * from './Table.type'
