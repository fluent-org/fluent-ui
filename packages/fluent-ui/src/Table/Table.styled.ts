import { styled, th, variant, css } from '../styles/styled'
import { TextAlign } from './Table.type'

export const StyledTableWrapper = styled.table`
  position: relative;
  overflow: hidden;
  display: table;
  border-spacing: 0;
  border-collapse: collapse;
  box-shadow: ${th.shadow('1')};
`

export const StyledTableHeadWrapper = styled.thead`
  display: table-header-group;
`
export const StyledTableBodyWrapper = styled.tbody`
  display: table-row-group;
`
export const StyledTableFooterWrapper = styled.tfoot`
  display: table-row-group;
`
export const StyledTableRowWrapper = styled.tr<{ isBodyRow: boolean }>`
  display: table-row;
  color: inherit;
  outline: none;
  &:hover {
    background-color: ${({ isBodyRow }): string => isBodyRow && th.color('standard.light2')};
  }
`

const textAlign = variant({
  prop: 'textAlign',
  default: 'inherit',
  variants: {
    inherit: css`
      text-align: inherit;
    `,
    left: css`
      text-align: left;
    `,
    center: css`
      text-align: center;
    `,
    right: css`
      text-align: right;
    `,
    justify: css`
      text-align: justify;
    `
  }
})
export const StyledTableHeadCellWrapper = styled.th<{ textAlign: TextAlign }>`
  display: table-cell;
  padding: 14px 40px 14px 16px;
  vertical-align: middle;
  font-size: 0.75rem;
  font-weight: 500;
  line-height: 1.3;
  letter-spacing: 0.015em;
  border-bottom: 1px solid ${th.color('standard.default')};
  ${textAlign}
`
export const StyledTableCellWrapper = styled.td<{ textAlign: TextAlign }>`
  display: table-cell;
  padding: 14px 40px 14px 16px;
  vertical-align: middle;
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.5;
  letter-spacing: 0.015em;
  border-bottom: 1px solid ${th.color('standard.default')};
  ${textAlign}
`
