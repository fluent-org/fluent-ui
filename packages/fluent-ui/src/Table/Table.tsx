import * as React from 'react'
import { StyledTableWrapper } from './Table.styled'
import { TableProps, ColumnsType, DataType, TablePropTypes } from './Table.type'
import Head from './components/Head'
import Body from './components/Body'
import Row from './components/Row'
import Cell from './components/Cell'
import Footer from './components/Footer'

function getColumnsChildren<T extends DataType>(
  columns: ColumnsType<T>[],
  data: T[]
): React.ReactFragment {
  const theHeadChildren = columns.map(
    (cell): React.ReactElement => {
      return <Cell key={cell.field}>{cell.title}</Cell>
    }
  )
  const theBodyChildren = data.map(
    (row, index): React.ReactElement => (
      <Row key={row.key}>
        {columns.map(
          (cell): React.ReactElement => {
            const text = row[cell.field]
            const content = typeof cell.render === 'function' ? cell.render(text, row, index) : text
            return <Cell key={text}>{content}</Cell>
          }
        )}
      </Row>
    )
  )
  return (
    <>
      <Head>
        <Row>{theHeadChildren}</Row>
      </Head>
      <Body>{theBodyChildren}</Body>
    </>
  )
}

class Table<T extends DataType> extends React.Component<TableProps<T>> {
  public static displayName = 'FTable'

  public static propTypes = TablePropTypes

  public static Head = Head
  public static Body = Body
  public static Footer = Footer
  public static Row = Row
  public static Cell = Cell

  private get theColumnsChildren(): React.ReactFragment {
    const { data, columns }: TableProps<T> = this.props
    return getColumnsChildren<T>(columns as ColumnsType<T>[], data as T[])
  }

  public render(): React.ReactElement {
    const { columns, children, ...rest }: TableProps<T> = this.props
    const theChildren = columns ? this.theColumnsChildren : children
    return <StyledTableWrapper {...rest}>{theChildren}</StyledTableWrapper>
  }
}

export default Table
