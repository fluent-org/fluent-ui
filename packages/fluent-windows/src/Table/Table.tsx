import * as React from 'react'
import classNames from 'classnames'
import { createUseStyles } from '@fluent-windows/styles'
import { Theme } from '../styles'
import { styles } from './Table.styled'
import { TableProps, ColumnsType, DataType, TablePropTypes, TableClassProps } from './Table.type'

import Head from './components/Head'
import Body from './components/Body'
import Row from './components/Row'
import Cell from './components/Cell'

function getColumnsChildren<T extends DataType>(
  columns: ColumnsType<T>[],
  data: T[]
): React.ReactFragment {
  const theHeadChildren = columns.map(
    (cell): React.ReactElement => {
      return <Cell key={cell.field}>{cell.title}</Cell>
    }
  )
  const theBodyChildren = data.map(
    (row, index): React.ReactElement => (
      <Row key={row.key}>
        {columns.map(
          (cell): React.ReactElement => {
            const text = row[cell.field]
            const content = typeof cell.render === 'function' ? cell.render(text, row, index) : text
            return <Cell key={text}>{content}</Cell>
          }
        )}
      </Row>
    )
  )
  return (
    <>
      <Head>
        <Row>{theHeadChildren}</Row>
      </Head>
      <Body>{theBodyChildren}</Body>
    </>
  )
}

export const name = 'Table'

const useStyles = createUseStyles<Theme, TableClassProps>(styles, { name })

function Table<T extends DataType>(props: TableProps<T>): React.ReactElement {
  const { className: classNameProp, data, columns, children, ...rest }: TableProps<T> = props
  const theChildren = columns
    ? getColumnsChildren<T>(columns as ColumnsType<T>[], data as T[])
    : children

  const classes = useStyles(props)
  const className = classNames(classes.root, classNameProp)
  return (
    <table className={className} {...rest}>
      {theChildren}
    </table>
  )
}

Table.displayName = `F${name}`

Table.propTypes = TablePropTypes

export default Table
