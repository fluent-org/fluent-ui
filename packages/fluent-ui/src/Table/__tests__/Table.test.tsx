import * as React from 'react'
import * as renderer from 'react-test-renderer'
import 'jest-styled-components'

import { ThemeProvider, Table } from '../..'
import { ColumnsType, DataType } from '../Table.type'

interface Data extends DataType {
  name: React.ReactText
  surname: React.ReactText
  birthYear: React.ReactText
  birthCity: React.ReactText
}

describe('Table', (): void => {
  const theme = {}
  const data: Data[] = [
    { key: '1', name: 'Mehmet', surname: 'Baran', birthYear: 1987, birthCity: 22 },
    { key: '2', name: 'John', surname: 'Brown', birthYear: 1977, birthCity: 32 },
    { key: '3', name: 'Jim', surname: 'Black', birthYear: 1997, birthCity: 14 }
  ]

  const columns: ColumnsType<Data>[] = [
    {
      title: 'Name',
      field: 'name',
      render: (text: React.ReactText): React.ReactElement => <a href="#">{text}</a> // eslint-disable-line
    },
    { title: 'Surname', field: 'surname' },
    { title: 'BirthYear', field: 'birthYear' },
    {
      title: 'BirthCity',
      field: 'birthCity'
    }
  ]
  test('basic', (): void => {
    const component = renderer.create(
      <ThemeProvider theme={theme}>
        <Table>
          <Table.Head>
            <Table.Row>
              <Table.Cell>Name</Table.Cell>
              <Table.Cell>Surname</Table.Cell>
              <Table.Cell>BirthYear</Table.Cell>
              <Table.Cell>BirthCity</Table.Cell>
            </Table.Row>
          </Table.Head>
          <Table.Body>
            {data.map(
              (row): React.ReactElement => (
                <Table.Row key={row.key}>
                  <Table.Cell>{row.name}</Table.Cell>
                  <Table.Cell>{row.surname}</Table.Cell>
                  <Table.Cell>{row.birthYear}</Table.Cell>
                  <Table.Cell>{row.birthCity}</Table.Cell>
                </Table.Row>
              )
            )}
          </Table.Body>
        </Table>
      </ThemeProvider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('prop column', (): void => {
    const component = renderer.create(
      <ThemeProvider theme={theme}>
        <Table<Data> columns={columns} data={data} />
      </ThemeProvider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
