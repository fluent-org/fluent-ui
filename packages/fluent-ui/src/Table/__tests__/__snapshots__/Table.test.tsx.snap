// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table basic 1`] = `
.c0 {
  position: relative;
  overflow: hidden;
  display: table;
  border-spacing: 0;
  border-collapse: collapse;
  box-shadow: 0px 1.6px 3.6px 0px rgba(0,0,0,0.132),0px 0.3px 0.9px 0px rgba(0,0,0,0.108);
}

.c1 {
  display: table-header-group;
}

.c4 {
  display: table-row-group;
}

.c2 {
  display: table-row;
  color: inherit;
  outline: none;
}

.c5 {
  display: table-row;
  color: inherit;
  outline: none;
}

.c5:hover {
  background-color: #f5f5f5;
}

.c3 {
  display: table-cell;
  padding: 14px 40px 14px 16px;
  vertical-align: middle;
  font-size: 0.75rem;
  font-weight: 500;
  line-height: 1.3;
  -webkit-letter-spacing: 0.015em;
  -moz-letter-spacing: 0.015em;
  -ms-letter-spacing: 0.015em;
  letter-spacing: 0.015em;
  border-bottom: 1px solid #dedede;
  text-align: left;
}

.c6 {
  display: table-cell;
  padding: 14px 40px 14px 16px;
  vertical-align: middle;
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.5;
  -webkit-letter-spacing: 0.015em;
  -moz-letter-spacing: 0.015em;
  -ms-letter-spacing: 0.015em;
  letter-spacing: 0.015em;
  border-bottom: 1px solid #dedede;
  text-align: left;
}

<table
  className="c0"
>
  <thead
    className="c1"
  >
    <tr
      className="c2"
    >
      <th
        className="c3"
      >
        Name
      </th>
      <th
        className="c3"
      >
        Surname
      </th>
      <th
        className="c3"
      >
        BirthYear
      </th>
      <th
        className="c3"
      >
        BirthCity
      </th>
    </tr>
  </thead>
  <tbody
    className="c4"
  >
    <tr
      className="c5"
    >
      <td
        className="c6"
      >
        Mehmet
      </td>
      <td
        className="c6"
      >
        Baran
      </td>
      <td
        className="c6"
      >
        1987
      </td>
      <td
        className="c6"
      >
        22
      </td>
    </tr>
    <tr
      className="c5"
    >
      <td
        className="c6"
      >
        John
      </td>
      <td
        className="c6"
      >
        Brown
      </td>
      <td
        className="c6"
      >
        1977
      </td>
      <td
        className="c6"
      >
        32
      </td>
    </tr>
    <tr
      className="c5"
    >
      <td
        className="c6"
      >
        Jim
      </td>
      <td
        className="c6"
      >
        Black
      </td>
      <td
        className="c6"
      >
        1997
      </td>
      <td
        className="c6"
      >
        14
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table prop column 1`] = `
.c0 {
  position: relative;
  overflow: hidden;
  display: table;
  border-spacing: 0;
  border-collapse: collapse;
  box-shadow: 0px 1.6px 3.6px 0px rgba(0,0,0,0.132),0px 0.3px 0.9px 0px rgba(0,0,0,0.108);
}

.c1 {
  display: table-header-group;
}

.c4 {
  display: table-row-group;
}

.c2 {
  display: table-row;
  color: inherit;
  outline: none;
}

.c5 {
  display: table-row;
  color: inherit;
  outline: none;
}

.c5:hover {
  background-color: #f5f5f5;
}

.c3 {
  display: table-cell;
  padding: 14px 40px 14px 16px;
  vertical-align: middle;
  font-size: 0.75rem;
  font-weight: 500;
  line-height: 1.3;
  -webkit-letter-spacing: 0.015em;
  -moz-letter-spacing: 0.015em;
  -ms-letter-spacing: 0.015em;
  letter-spacing: 0.015em;
  border-bottom: 1px solid #dedede;
  text-align: left;
}

.c6 {
  display: table-cell;
  padding: 14px 40px 14px 16px;
  vertical-align: middle;
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.5;
  -webkit-letter-spacing: 0.015em;
  -moz-letter-spacing: 0.015em;
  -ms-letter-spacing: 0.015em;
  letter-spacing: 0.015em;
  border-bottom: 1px solid #dedede;
  text-align: left;
}

<table
  className="c0"
  data={
    Array [
      Object {
        "birthCity": 22,
        "birthYear": 1987,
        "key": "1",
        "name": "Mehmet",
        "surname": "Baran",
      },
      Object {
        "birthCity": 32,
        "birthYear": 1977,
        "key": "2",
        "name": "John",
        "surname": "Brown",
      },
      Object {
        "birthCity": 14,
        "birthYear": 1997,
        "key": "3",
        "name": "Jim",
        "surname": "Black",
      },
    ]
  }
>
  <thead
    className="c1"
  >
    <tr
      className="c2"
    >
      <th
        className="c3"
      >
        Name
      </th>
      <th
        className="c3"
      >
        Surname
      </th>
      <th
        className="c3"
      >
        BirthYear
      </th>
      <th
        className="c3"
      >
        BirthCity
      </th>
    </tr>
  </thead>
  <tbody
    className="c4"
  >
    <tr
      className="c5"
    >
      <td
        className="c6"
      >
        <a
          href="#"
        >
          Mehmet
        </a>
      </td>
      <td
        className="c6"
      >
        Baran
      </td>
      <td
        className="c6"
      >
        1987
      </td>
      <td
        className="c6"
      >
        22
      </td>
    </tr>
    <tr
      className="c5"
    >
      <td
        className="c6"
      >
        <a
          href="#"
        >
          John
        </a>
      </td>
      <td
        className="c6"
      >
        Brown
      </td>
      <td
        className="c6"
      >
        1977
      </td>
      <td
        className="c6"
      >
        32
      </td>
    </tr>
    <tr
      className="c5"
    >
      <td
        className="c6"
      >
        <a
          href="#"
        >
          Jim
        </a>
      </td>
      <td
        className="c6"
      >
        Black
      </td>
      <td
        className="c6"
      >
        1997
      </td>
      <td
        className="c6"
      >
        14
      </td>
    </tr>
  </tbody>
</table>
`;
